### ISSUE TITLE
Consider alternative name for "modules"

### ISSUE THREAD

**Issue by GShotwell** [rstudio/py-shiny#539]
We chose the name "modules" for two main reasons:

1) It's the same as the R term which makes transitioning a bit easier
2) Shiny modules are pretty similar to generic Python modules so it made sense to use the name

The name does seem to create confusion in some users and is difficult to explain. For example I ended up saying: "the Shiny `module_ui` function is in Shiny's UI module". 

According to chatGPT the most common term for this is "Component" so maybe we should go with `@component.ui` and `@component.server`.

Web Framework | Reusable Elements
-- | --
React | Components
Angular | Components
Vue.js | Components
Ember.js | Components
Backbone.js | Views or Templates
Meteor.js | Templates or Components
Express.js | Middleware or Router Middleware
Next.js | Components
Svelte | Components
Preact | Components
Django | Templates or Views
Dash | Components
Streamlit | Components or Widgets

**Comment by nealrichardson** [comment]
+1 for component over module, definitely seems closer to the right concept. Maybe a rename plus better positioning in the docs would be enough: for me, "modules" felt so far advanced into package development that I didn't even open the docs page to see that it actually was the solution for encapsulating a feature for reuse.

**Comment by vnijs** [comment]
Good suggestion. On a related note, I had hoped to be able to use classes and methods for reuse in py-shiny. This would seem particularly useful In a bigger apps where you may be able to inherit a lot from a class and only have to overwrite a few methods

I'm not sure if @wch's suggested at the end of issue [below](https://github.com/rstudio/py-shiny/issues/510) is what would be needed to make that work now or if this shows a best-case-scenario for the future.

**Comment by cpsievert** [comment]
Component doesn't seem right either since the way most projects use that term is closer to what Shiny would call a binding/widget than it is to the concept of a Shiny module. So, if we are going to change the name, I'd probably prefer something like "Template" or "Namespace".

**Comment by GShotwell** [comment]
@cpsievert I'm not sure that's right, for example if I'm creating a React component I encapsulate some React code into something which can be used by other parts of my application. This seems exactly the same as what we do in Shiny for Python. Similarly, [Solara](https://solara.dev/docs/fundamentals/components) and [Pynecone](https://pynecone.io/docs/components/overview) both use component to refer to what we would call modules.

Template to me sounds a bit more like a code snippet or HTML template which isn't quite right. Namespace is probably the most accurate, but it's not that evocative and we'd have the similar problem of being the only project that uses the term.

**Comment by nealrichardson** [comment]
I agree that "template" and "namespace" don't feel right, and also shared @vnijs's experience that what I was really looking for was "class" because this feels like a natural thing for OOP.
